File: java/src/Boggle.java
import java.util.*;

public class Boggle {

    public static void main(String[] args) {
        Boggle boggle = new Boggle();

        List<Character> board = Arrays.asList('G', 'I', 'Z', 'U', 'E', 'K', 'Q', 'S', 'E');
        int n = 3;
        int m = 3;
        Set<String> dict = new HashSet<>(Arrays.asList("GEEKS", "FOR", "QUIZ", "GO"));

        boggle.findWords(board, n, m, dict);
    }

    private boolean isWord(String str, Set<String> dict) {
        return dict.contains(str);
    }

    private void findWordsUtil(List<Character> board, int n, int m, List<Boolean> visited,
                               int i, int j, StringBuffer str, Set<String> dict) {

        visited.set(i * m + j, true);
        str.append(board.get(i * m + j));

        if (isWord(str.toString(), dict)) {
            System.out.println(str.toString());
        }

        for (int row = i-1; row <= i+1 && row < n; row++) {
            for (int col = j-1; col <= j+1 && col < m; col++) {
                if (row >= 0 && col >= 0 && !visited.get(row * m + col)) {
                    findWordsUtil(board, n, m, visited, row, col, str,  dict);
                }
            }
        }

        str.deleteCharAt(str.length() - 1);
        visited.set(i * m + j, false);
    }

    private void findWords(List<Character> board, int n, int m, Set<String> dict) {
        List<Boolean> visited = new ArrayList<>();
        for (int i = 0; i < n * m; i++) {
            visited.add(false);
        }

        StringBuffer stringBuffer = new StringBuffer();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                findWordsUtil(board, n, m, visited, i, j, stringBuffer, dict);
            }
        }
    }
}

File: python/boggle.py
def find_words_util(board, dictionary, i, j, n, m, visited, letters):
    letters.append(board[i][j])
    word = ''.join(letters)
    if word in dictionary:
        print(word)
    
    visited[i][j] = True
    for k in range(i - 1, i + 2):
        for l in range(j - 1, j + 2):
            if k >= 0 and k < n and l >= 0 and l < m and not visited[k][l]:
                find_words_util(board, dictionary, k, l, n, m, visited, letters)
    
    del letters[-1]
    visited[i][j] = False


def find_words(board, dictionary, n, m):
    visited = []
    for i in range(n):
        visited.append([])
        for j in range(m):
            visited[i].append(False)

    for i in range(n):
        for j in range(m):
            find_words_util(board, dictionary, i, j, n, m, visited, [])




board = [['G', 'I', 'Z'], ['U', 'E', 'K'], ['Q', 'S', 'E']]
dictionary = set(['GEEKS', 'FOR', 'QUIZ', 'GO'])

find_words(board, dictionary, 3, 3)